src/rocksdb/src/lib.rs

```diff
@@ -0,0 +1,514 @@
+                        }
+                        None => writes.delete(encoded_key.as_slice()),
+                    }
+                    let previous_value = match previous_value.map(|v| V::decode(&v)).transpose() {
+                        Ok(v) => v,
+                        Err(e) => {
+                            let _ = response_sender.send(Err(Error::DecodeError(e)));
+                            return;
+                        }
+                    };
+
+                    result.push(UpsertResult {
+                        key,
+                        val,
+                        previous_value,
+                    });
+                }
+                // Perform the multi_get and record metrics, if there wasn't an error.
+                let now = Instant::now();
+                match db.write_opt(writes, &wo) {
+                    Ok(()) => {
+                        let latency = now.elapsed().as_micros();
+                        metrics.write_latency_us.observe(f64::cast_lossy(latency));
+                        metrics
+                            .write_batch_size
+                            .observe(f64::cast_lossy(batch_size));
+                    }
+                    Err(e) => {
+                        let _ = response_sender.send(Err(Error::RocksDB(e)));
+                        return;
+                    }
+                }
+                // debug!("finished writing batch size({size}) for worker {worker_id}, source {source_id}");
+
+                let _ = response_sender.send(Ok(result));
+            }
+
+            // Gracefully cleanup if the `RocksDBInstance` has gone away.
+            db.cancel_all_background_work(true);
+            drop(db);
+        });
+
+        Ok(Self { tx })
+    }
+
+    /// Upsert a batch of values, returning the keys and values back, along
+    /// with previous values. If this errors, the `RocksDB` instance becomes
+    /// unusable.
+    pub async fn upsert(
+        &mut self,
+        batch: Vec<UpsertValue<K, V>>,
+    ) -> Result<Vec<UpsertResult<K, V>>, Error> {
+        if batch.is_empty() {
+            return Ok(Vec::new());
+        }
+
+        let (tx, rx) = oneshot_channel();
+        self.tx
+            .send(Command::Upsert {
+                batch,
+                response_sender: tx,
+            })
+            .await
+            .map_err(|_| Error::RocksDBThreadGoneAway)?;
+
+        // We also unwrap all rocksdb errors here.
+        rx.await.map_err(|_| Error::RocksDBThreadGoneAway)?
+    }
+
+    /// Gracefully shutdown RocksDB. Can error if the instance
+    /// is already shut down or errored.
+    pub async fn close(self) -> Result<(), Error> {
+        let (tx, rx) = oneshot_channel();
+        self.tx
+            .send(Command::Shutdown { done_sender: tx })
+            .await
+            .map_err(|_| Error::RocksDBThreadGoneAway)?;
+
+        let _ = rx.await;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
```

Comment:
These tests look like they're meant to test the public API of the crate, so would be better as "integration" tests in tests/rocksdb.rs.

Reply:
I typically agree with @danhhz, but I moved it anyways, and found some issues with publicity of fields, so ill leave it

Reply:
> IANGus, but integration tests just mean something else to link and IMO the benefits of being restricted to only seeing the public API in tests are marginal. so better to keep anything that can be a unit test as a unit test

FWIW, I try to keep it to a single integration test binary (possibly split across multiple files), a la

```
tests/integration.rs
tests/integration/file1.rs
tests/integration/file2.rs
```

to avoid the annoying behavior of getting one integration test binary per top-level file.

But, yeah, in the past when I've written crates in isolation, writing unit tests rather than integration tests has basically guaranteed that I forgot to mark some crucial API as `pub`. Not a huge deal to fix up a few missing `pub`s in post, but I personally like defaulting to testing the external interface, and only testing the internal interface when required.